### PR TITLE
.git: add codespell workflow

### PR DESCRIPTION
**Pull request by tchaikov**
to identify misspelling in the code.

The GitHub actions in this workflow run codespell when a new pull
request is created targetting master or enterprise branch. Errors
will be annotated in the pull request. A new entry along with the
existing tests like build, unit test and dtest will be added to the
"checks" shown in github PR web UI. one can follow the "Details" to
find the details of the errors.

unfortunately, this check checks all text files unless they
are explicitly skipped, not just the new ones added / changed in the
PR under test. in other words, if there are 42 misspelling
errors in master, and you are adding a new one in your PR,
this workflow shows all of the 43 errors -- both the old
and new ones.

the misspelling in the code hurts the user experience and some
time developer's experience, but the text files under test/cql
can be sensitive to the text, sometimes, a tiny editing could
break the test, so it is added to the skip list.

So far, we don't consider the errors are just represented for
information purpose, so even if the check fails, it is not considered
as blockers for merging the tested PR. we can change this in github's
Branch protectionn rule on a per-branch basis.

### CODE REVIEW

File: .github/workflows/codespell.yaml
Comment (mykaul) [inline .github/workflows/codespell.yaml:19]: YES! Thank you, it was driving me nuts ;-)

File: .github/workflows/codespell.yaml
Comment (mykaul) [inline .github/workflows/codespell.yaml:20]: I believe we need to ignore Seastar as well (generally, all submodules) in Scylla (we should / can fix them in their original repos?)

**Comment by mykaul** [comment]
VERY nice!

File: .github/workflows/codespell.yaml
Comment (tchaikov) [reply .github/workflows/codespell.yaml:20]: see https://github.com/actions/checkout/blob/main/README.md#usage and https://github.com/actions/checkout/blob/b4ffde65f46336ab88eb53be808477a3936bae11/action.yml#L82-L90 . 
i am quoting the document here as well:
```yaml
    # Whether to checkout submodules: `true` to checkout submodules or `recursive` to
    # recursively checkout submodules.
    #
    # When the `ssh-key` input is not provided, SSH URLs beginning with
    # `git@github.com:` are converted to HTTPS.
    #
    # Default: false
    submodules: ''
```
by default, `submodules` is false. so the submodules are not fetched.

File: .github/workflows/codespell.yaml
Comment (nyh) [inline .github/workflows/codespell.yaml:6]: Would have been nice if, for reviewers like me who are not proficient in github workflows, you could summarize what this PR would do if committed. E.g., do these lines say that codespell will be run automatically on every PR? Where will the result be visible, and will it be informational or will block the PR if mispellings are found?

What does it actually check - the new text in the PR, or the entire project again and again? If it's the latter, why is interesting to see 100 pre-existing spelling errors again and again every time you send a PR?

I think all of this should be explained in the commit message.

File: .github/workflows/codespell.yaml
Comment (nyh) [reply .github/workflows/codespell.yaml:18]: Why do we have this word "ser" in our documents enough to drive you nuts?

**Comment by tchaikov** [comment]
v2:

- change `fetch-depth` from 0 to "1" which is the default value, as we don't need to fetch all the history.

File: .github/workflows/codespell.yaml
Comment (nyh) [reply .github/workflows/codespell.yaml:19]: Another directory you should maybe consider excluding is test/cql. These look like text files, but really aren't, and "fixing" words there can break the tests. Although usually, if you fix both the .cql and .result file in the same way, it should be fine.

**Comment by mykaul** [comment]
Documentation for this GH workflow plugin is @ https://github.com/codespell-project/actions-codespell#usage

**Comment by nyh** [comment]
> Documentation for this GH workflow plugin is @ https://github.com/codespell-project/actions-codespell#usage

I can also use Google to find documentation. But I prefer a commit message that is stand-alone and explains exactly what happens after a patch (see my specific questions above), without requiring me to guess or spend hours reading documentation.

**Comment by mykaul** [comment]
> > Documentation for this GH workflow plugin is @ https://github.com/codespell-project/actions-codespell#usage
> 
> I can also use Google to find documentation. But I prefer a commit message that is stand-alone and explains exactly what happens after a patch (see my specific questions above), without requiring me to guess or spend hours reading documentation.

It could have been added to the commit message, agreed. Since it wasn't, I'm adding it here.

**Comment by mykaul** [comment]
> > @nyh codespell always suggest 'ser' is misspelled for 'set', and we use 'ser' (for service) quite a lot...
> 
> I still don't understand. I thought you said we only use codespell to check documentation - external and internal (i.e., comments). The PR title also has the prefix "docs". Why would we use the word "ser" as a shortcut for "service" in documentation?

I think it should be re-worded - it should not be just for documentation.

File: .github/workflows/codespell.yaml
Comment (tchaikov) [reply .github/workflows/codespell.yaml]: thanks. added.

File: .github/workflows/codespell.yaml
Comment (tchaikov) [reply .github/workflows/codespell.yaml:6]: @nyh i amended the commit message. does it address your questions now?

**Comment by benipeled** [comment]
I have no problem with the workflow syntax itself but two comments
1. If there are - please make sure to fix all codespell warns/errors before merging this change - otherwise people will get used to it and ignore it
1. I'd add it to Jenkins (just because we agreed a long time ago to implement scylla CI on Jenkins) but since you already add it and verify we can give it a chance (theoretically there should be no issue to run on different platforms for this task)

**Comment by nyh** [comment]
>     1. If there are - please make sure to fix all codespell warns/errors before merging this change - otherwise people will get used to it and ignore it

I thought @tchaikov said above that it only checks the PR, not all the pre-existing code? @tchaikov please clarify.

**Comment by tchaikov** [comment]
> I have no problem with the workflow syntax itself but two comments
> 
>     1. If there are - please make sure to fix all codespell warns/errors before merging this change - otherwise people will get used to it and ignore it
> 

please consider these two options: a green check result, with 100 error in the "details" link. or a red check result, but nobody cares about it. personally, i think the latter is worse.

>     2. I'd add it to Jenkins (just because we agreed a long time ago to implement scylla CI on Jenkins) but since you already add it and verify we can give it a chance (theoretically there should be no issue to run on different platforms for this task)

i don't feel strong either way. if some of us has more bandwidth on creating this jenkins job, i am happy to close this PR or revert it if it gets merged.

**Comment by tchaikov** [comment]
> > ```
> > 1. If there are - please make sure to fix all codespell warns/errors before merging this change - otherwise people will get used to it and ignore it
> > ```
> 

and i don't have enough bandwidth to fix all of them at this moment. i'd fix them in piece meal, as it would need more much efforts. fortunately, Yaniv's working on this. i think this PR could serve as a metric for our efforts in this direction. 

> I thought @tchaikov said above that it only checks the PR, not all the pre-existing code? @tchaikov please clarify.

no, it checks all of them. i revised the commit message to clarify this.

**Comment by tchaikov** [comment]
v3:

- further clarify that the tool checks all spelling problems - both the existing ones and new ones.
- add "only_warn: 1", but i am not in favor of covering errors. i don't think the brave developers / maintainers would be blocked by red crosses as long as they are not fatal. =)

**Comment by tchaikov** [comment]
well, the test is green. but it does not mean anything =(

**Comment by benipeled** [comment]
> > I have no problem with the workflow syntax itself but two comments
> > ```
> > 1. If there are - please make sure to fix all codespell warns/errors before merging this change - otherwise people will get used to it and ignore it
> > ```
> 
> please consider these two options: a green check result, with 100 error in the "details" link. or a red check result, but nobody cares about it. personally, i think the latter is worse.

Me2, and it make sense if you don't have the time to deal with them all now - we just need to make sure it will be solved or added to ignore list, otherwise it will ignore

> 
> > ```
> > 2. I'd add it to Jenkins (just because we agreed a long time ago to implement scylla CI on Jenkins) but since you already add it and verify we can give it a chance (theoretically there should be no issue to run on different platforms for this task)
> > ```
> 
> i don't feel strong either way. if some of us has more bandwidth on creating this jenkins job, i am happy to close this PR or revert it if it gets merged.

Since you already did it - we'll move on with your solution - and thanks for that!

**Comment by mykaul** [comment]
https://github.com/scylladb/scylladb/pull/16289 should take care of few more typos in our codebase.

**Comment by nyh** [comment]
> no, it checks all of them. i revised the commit message to clarify this.

Thanks for revising the commit message.
I'm not sure how this will be useful if we always have this "red x" and everyone will ignore it, but if it's not **mandatory** - and maintainers are allowed to ignore it - I guess we can start with that, perhaps as a tool for people like @mykaul who are trying to get rid of the pre-existing errors.

**Comment by nyh** [comment]
> I've fixed some, I have yet another in-the-works PR to fix some more, which will reduce the number significantly. I'm not sure it'll go down to 0, but we'll have to add some 'ignore' words, I reckon.

I'm worried that it's not just ignore words... For example, in test/cql-pytest/test_uda.py in commit 48bd7529711 @wmitros used a variable named "siz", supposedly as a cute shortcut for "size". Should this be forbidden and we should change the variable name to "size"? Or will we add "siz" to the list of ignore words? Neither of these options seem great to me... Outright spelling mistakes should be forbidden in function names, but we don't really need a rule against silly shorthand. 

On the other hand, we don't want the word "siz" inside a user-facing document to also be accepted. So I don't know what to suggest. I wonder if we can configure codespell to ignore variable names shorter than, say, 4 characters, so variable names like "siz", "x" etc., will not look like errors, but don't have such an exception in real text (documentations, comments, etc.)?

**Comment by tchaikov** [comment]
if we check the annotations, we can found that github does differentiate the "Unchanged files with check annotations" from (presumably) changed file with check annotations. see https://github.com/scylladb/scylladb/pull/16285/files

so this tool helps reviewer and contributor to check the changed part and to ignore the new ones if they believe they are benign. it's not mandatory, but it's right there if we want to check it.

File: .github/workflows/codespell.yaml
Comment (nyh) [reply .github/workflows/codespell.yaml:6]: Yes

**Comment by tchaikov** [comment]
v4:

- add "only_warn: 1". the error is distracting. after a second thought, i think we should make the check pass even there are misspellings -- we can always make it more scaring when we are ready for it.

**Comment by tchaikov** [comment]
@benipeled hi Beni, i think i've addressed your comment. could you take another look?

**Comment by benipeled** [comment]
> @benipeled hi Beni, i think i've addressed your comment. could you take another look?

Thanks @tchaikov, lgtm

**Review by mykaul — APPROVED** [review]

**Comment by tchaikov** [comment]
@scylladb/scylla-maint hello maintainers, could you help queue this change?

**Comment by mykaul** [comment]
> @scylladb/scylla-maint hello maintainers, could you help queue this change?

Last batch, perhaps before we add this? https://github.com/scylladb/scylladb/pull/16388

**Comment by denesb** [comment]
> > @scylladb/scylla-maint hello maintainers, could you help queue this change?
> 
> Last batch, perhaps before we add this? #16388

I queued #16388, will queue this after promotion, to minimize noise, when people have their first impression of this new check.

**Comment by tchaikov** [comment]
thanks. but actually, this change just adds a new green check even if fails.

**Comment by denesb** [comment]
> thanks. but actually, this change just adds a new green check even if it fails.

Ok, queuing then.

**Pull request closed without merging**
